CER-65-Permit OnBehalfOfAccount

andrew-certora opened the pull request:
Implements CER-65: https://linear.app/euler-labs/issue/CER-65/permit-onbehalfofaccount

Previously this PR included a start to CER-16 https://linear.app/euler-labs/issue/CER-16/permit-authorization-rules
but that one will be harder than anticipated so is moved to this other PR: https://github.com/euler-xyz/ethereum-vault-connector/pull/103 so as not to block CER-65


Also includes a small refactor of `actualCaller`

kasperpawlowski commented on certora/specs/CER-15-Permit/CER-65-Permit-onBehalfOfAccount.spec:
```diff
@@ -0,0 +1,53 @@
+// with onBehalfOfAccount == victim in the execution context,
+// and an adversary attempts to execute a call with a different
+// onbehalf of address, the call will revert
+rule cannot_spoof_onBehalfOfAccount_call() {
```
it seems that the only reason why this holds is this line of code:
https://github.com/euler-xyz/ethereum-vault-connector/blob/644a61f687d724a5070e7070e6054782023766b0/src/EthereumVaultConnector.sol#L767

wouldn't it be better to prove it as follows?
- have a hook on `ExecutionContext.getOnBehalfOfAccount`
- filter out `getCurrentOnBehalfOfAccount`
- acknowledge that `msg.sender == EVC` only in the `permit` context as per `CER-51`
- assert in the hook that `ExecutionContext.getOnBehalfOfAccount` is only called when `msg.sender == EVC`

note that what we want to prove is that we only rely on the stored address value if in `permit` context. so that even if the address spoofed, we'll never rely on it unless in the `permit` which will get further verified as per `CER-18`, `CER-19` and `CER-72`

andrew-certora replied:
Ah Okay. I did misunderstand what you meant. I thought by " EVC MUST only rely on the stored Execution Context's onBehalfOfAccount address when in Permit context" you meant that within a permit, only onBehalfOfAccount from the context (rather than arguments from call/batch) is used to determine the relevant address, but actually you meant that permit is the only place where this will be used.

I am working on re-specifying this.

kasperpawlowski replied:
sorry for the confusion!

andrew-certora replied:
No worries. I just pushed another version that I think captures what you had in mind. Note that we have to use a funciton summary rather than a hook to latch on to the contexts in which `getOnBehalfOfAccount` is accessed since it is a function. It is verified: https://prover.certora.com/output/65266/89eb3cb2b6ea4ce0a0abe96576da7b83?anonymousKey=2208f43b6fce8ff1a24d4995cb3e366cd900c18e

By the way, what would you like me to do with these other 2 rules I initially wrote for CER-65. I can either:
* get rid of them if they are not useful/interesting
* keep them with the Permit specs, but make a new linear ticket and give them the new number
* keep them but move them into misc
(I shouldn't leave these in this file because even though they still pass for now, in principle the newly introduced function summary may interfere with them if there are code or spec changes later)

andrew-certora commented on certora/specs/CER-15-Permit/CER-16-Permit-authorization.spec:
```diff
@@ -0,0 +1,43 @@
+    // We use the signer from the permit call as the actualCaller
+    // for the call.
+    require actualCaller(e) == signer;
+    require e.msg.sender == currentContract;
```
I think you are right. That is better. This one is a bit tricky, so I might need a bit more time to fix it. CER-16 is now WIP so no need to re-review it yet.

andrew-certora commented:
I moved CER-16 out of this PR to a new one: https://github.com/euler-xyz/ethereum-vault-connector/pull/103
to unblock CER-65. I updated the description.

andrew-certora commented:
(the rereview is just for the merge conflict resolution which required adding a curly brace)

andrew-certora commented on certora/conf/CER-15-Permit/CER-65-Permit-onBehalfOfAccount.conf:
```diff
@@ -0,0 +1,18 @@
+{
+    "files": [
+        "src/ExecutionContext.sol",
```
fixed

andrew-certora commented on certora/harness/EthereumVaultConnectorHarness.sol:
```diff
@@ -40,6 +40,15 @@ contract EthereumVaultConnectorHarness is EthereumVaultConnector {
         return operatorLookup[addressPrefix][operator];
     }
+
+    function selfCallSuccessCheck(uint256 value, bytes calldata data) external returns (bool) {
```
fixed

andrew-certora commented:
Apologies. Needs re-review after rebase.

andrew-certora commented:
This just needs another post rebase review. I had a few in parallel.
